software/glasgow/applet/audio/dac/__init__.py

```diff
@@ -120,7 +136,52 @@ def elaborate(self, platform):
         return m
 
 
-class AudioDACApplet(GlasgowApplet):
+class AudioDACInterface:
+    def __init__(self, logger: logging.Logger, assembly: AbstractAssembly, *,
+                 pins: tuple[GlasgowPin], width: Literal[1, 2], signed: bool):
+        self._logger = logger
+        self._level  = logging.DEBUG if self._logger.name == __name__ else logging.TRACE
+
+        component = assembly.add_submodule(AudioDACComponent(assembly.add_port(pins, name="dac"),
+                                                             width=width, signed=signed))
+        # The buffer size should be limited to prevent the buffer size growing unboundedly
+        # 1e6 was ~arbitrarily chosen
+        self._pipe = assembly.add_out_pipe(component.i_stream, buffer_size=1e6)
+        self._sample_clock = assembly.add_clock_divisor(
+            component.sample_divisor,
+            ref_period=assembly.sys_clk_period,
+            # Drift of sampling clock is extremely bad, so ensure it only happens insofar as
+            # the oscillator on the board is imprecise, and with no additional error.
+            tolerance=0,
+            round_mode="nearest",
+            name="sampling"
+        )
+        self._modulation_clock = assembly.add_clock_divisor(
+            component.modulation_divisor,
+            ref_period=assembly.sys_clk_period,
+            name="modulation"
+        )
+
+    @property
+    def sample_clock(self) -> ClockDivisor:
+        """:py:`ClockDivisor` for setting the sample rate."""
+        return self._sample_clock
+
+    @property
+    def modulation_clock(self) -> ClockDivisor:
+        """:py:`ClockDivisor` for setting the modulation rate."""
+        return self._modulation_clock
+
+    async def write(self, pcm_data: bytes | bytearray | memoryview):
+        """Send samples to the DAC to be played in round robin order."""
+        await self._pipe.send(pcm_data)
+
+    async def flush(self):
+        """Force any pending writes to finish sending all data."""
```

Comment:
This isn't *wrong* per se but it doesn't really explain the semantics. In this particular configuration, what the method does is:


```suggestion
        """Ensure any past writes have reached the device FIFO."""
```

Reply:
Thanks, I wasn't satisfied by my wording there, but wasn't sure how to phrase it more clearly.